### PR TITLE
Hg38 masked

### PR DESCRIPTION
**Pull request by ViktorHy**
Big update: Going live monday 23/10

Stubs have been tested and passed, thanks @ryanjameskennedy 

Added a profile for masked hg38 reference. This update includes WGS references build with this new genome build for GATK and has passed validation. Future updates to GENS PoN needs to be done once this has been in production for a month or so

Tests needs to be passed:
Oncov1-0 sample []
Oncov2-0 sample []
WGS sample trio []
WGS sample single []

WGS is tested fully into Scout via validation and onco profiles does not need to be tested for scout import. Stubs could have added issues for onco. Run samples and make sure no crashes.

### CODE REVIEW

**Review by ryanjameskennedy — APPROVED** [review]
LGTM!

**Comment by ViktorHy** [comment]
the melt stub throws errors. Did you test that ryan?

File: main.nf [hunk 1]
```diff
@@ -363,6 +511,7 @@ process sentieon_qc {
 		set group, id, file("${id}_qc.json") into qc_cdm
 		set group, id, file("${id}_qc.json") into qc_melt
 		file("*.txt")
+		//path "*versions.yml"
```
Comment (Jakob37) [inline main.nf:514]: Should this one be commented here?

File: main.nf [hunk 2]
```diff
@@ -185,11 +187,32 @@ process fastp {
...
+		  container: ${task.container}
+		END_VERSIONS
+		"""
+
+	stub:
+		"""
+		touch ${id}_R1_a_q_u_trimmed.fq.gz
+		touch ${id}_R2_a_q_u_trimmed.fq.gz
+
+		cat <<-END_VERSIONS > ${task.process}_versions.yml
```
[the remark this answers is not in the document]
Comment (Jakob37) [reply main.nf:204]: The structure seems very similar for the different processes as well, just with different names and commands to extract versions. Could it be a tiny util function, to give the structure and extract process and container?

File: main.nf [hunk 3]
```diff
@@ -500,14 +728,15 @@ smn_tsv
...
+	publishDir "/${OUTDIR}/versions", mode: 'copy' , overwrite: 'true', pattern: '*versions.yml'
 	tag "$group"
 	cpus 2
 	time '10h'
 	memory '40 GB'
 	scratch true
 	stageInMode 'copy'
 	stageOutMode 'copy'
-	//publishDir "${OUTDIR}/plots/GAV/${group}", mode: 'copy' , overwrite: 'true', pattern: '*.png'
+	//publishDir "/${OUTDIR}/plots/GAV/${group}", mode: 'copy' , overwrite: 'true', pattern: '*.png'
```
Comment (Jakob37) [inline main.nf:739]: Path updated in comment. Should the comment be removed or included?

File: main.nf (hunk 2 above)
[the remark this answers is not in the document]
Comment (ViktorHy) [reply main.nf:204]: good question. I guess @ryan knows more about that. How does nf-core do it?

File: main.nf (hunk 1 above)
Comment (ViktorHy) [reply main.nf:514]: dont't know. I'm starting to regret adding the stubs into this PR

File: main.nf [hunk 4]
```diff
@@ -158,6 +158,7 @@ if(genome_file ){
 }
 
 process fastp {
+	publishDir "/${OUTDIR}/versions", mode: 'copy' , overwrite: 'true', pattern: '*versions.yml'
```
Comment (Jakob37) [inline main.nf:161]: Why is the `OUTDIR` prepended with "/"? If it is a "nextflow root" folder without the root slash, why not add the "/" to the variable itself?

File: main.nf [hunk 5]
```diff
@@ -413,16 +583,36 @@ process chanjo_sambamba {
 	output:
 		file("${id}.bwa.chanjo.cov") into chanjocov
```
[the remark this answers is not in the document]
Comment (ViktorHy) [reply main.nf:584]: good catch

File: main.nf (hunk 3 above)
Comment (ViktorHy) [reply main.nf:739]: some weird ass code. Don't even know what GAV is. Must have been something Ryan snuck in during testing

File: main.nf (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jakob37) [reply main.nf:204]: I checked the rare diseases. They seem to do it this way. But they have less content in each "version blob", only showing software names and versions.

File: main.nf [hunk 6]
```diff
@@ -1262,39 +2013,62 @@ process annotate_vep {
...
+			--force_overwrite \\
+			--fasta $params.VEP_FASTA \\
+			--dir_cache $params.VEP_CACHE \\
+			--dir_plugins $params.VEP_CACHE/Plugins \\
+			--distance 200 \\
+			-cache \\
+			--plugin CADD,$params.CADD \\
+			--plugin LoFtool \\
+			--plugin MaxEntScan,$params.MAXENTSCAN,SWA,NCSS \\
+			--plugin dbNSFP,/fs1/resources/ref/hg38/annotation_dbs/dbnsfp/dbNSFP4.3a_grch38.gz,transcript_match=1,REVEL_score,REVEL_rankscore \\
```
[the remark this answers is not in the document]
Comment (ViktorHy) [reply main.nf:2039]: maybe, it is what it is right now

File: main.nf (hunk 4 above)
Comment (ViktorHy) [reply main.nf:161]: extra precaution, never hurts to have an extra slash

**Review by Jakob37 — COMMENTED** [review]
Nice work!

Some thoughts after some initial checking. I feel a bit self-aware here leaving comments without having worked much in this code. Just let me know if I am "wrong on the ball" on any of the points.

* If I understand things correctly, the `output "*versions.yml"` and the `publishDir ...versions` should both be present for all processes which output versions? I found a bunch where this was not the case (see comments). There are more beyond these. My guess is that this needs to be fixed before this is merged.
* Is it fine to not version the Perl-scripts? (I realize they are present in the repo, so in a sense you have the version through the commit. But would still be useful to see whether they have changed between different runs.

I attempted to trace the masked params through the code. I didn't see anything strange there. But is there some particular part of the code that should be reviewed extra carefully for these updates? I think you could select ranges in the code and comment on them then, if you want more eyes on some sections.

File: main.nf (hunk 4 above)
Comment (Jakob37) [reply main.nf:161]: I see! What about this then:

```
OUTDIR = '/' + params.outdir+'/'+params.subdir
```

File: main.nf (hunk 1 above)
Comment (Jakob37) [reply main.nf:514]: I agree, it is hard to see what are critical changes in the diff

**Comment by ViktorHy** [comment]
I might just revert the stubs/versions from this PR. The code just breaks down. And processes are obviously not all finished for a release

**Comment by ViktorHy** [comment]
> Nice work!
> 
> Some thoughts after some initial checking. I feel a bit self-aware here leaving comments without having worked much in this code. Just let me know if I am "wrong on the ball" on any of the points.
> 
>     * If I understand things correctly, the `output "*versions.yml"` and the `publishDir ...versions` should both be present for all processes which output versions? I found a bunch where this was not the case (see comments). There are more beyond these. My guess is that this needs to be fixed before this is merged.
> 
>     * Is it fine to not version the Perl-scripts? (I realize they are present in the repo, so in a sense you have the version through the commit. But would still be useful to see whether they have changed between different runs.
> 
> 
> I attempted to trace the masked params through the code. I didn't see anything strange there. But is there some particular part of the code that should be reviewed extra carefully for these updates? I think you could select ranges in the code and comment on them then, if you want more eyes on some sections.

Perl-skripts are versioned with the versions of the nextflow code base. As they are only ever run with this code base. This something we do to reduce the amount of documentation and need for validation. Github releases cover them

**Comment by Jakob37** [comment]
> Perl-skripts are versioned with the versions of the nextflow code base. As they are only ever run with this code base. This something we do to reduce the amount of documentation and need for validation. Github releases cover them

OK, sounds reasonable!

File: main.nf (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanjameskennedy) [reply main.nf:204]: I actually like this idea but nf-core does it the way I've done it. We slightly modified it for @ramsainanduri's Read the Docs.

File: main.nf (hunk 3 above)
Comment (ryanjameskennedy) [reply main.nf:739]: No I only added `versions` directories. That was there before, I just changed all `"${OUTDIR}` to `/"${OUTDIR}`.

File: main.nf (hunk 1 above)
Comment (ryanjameskennedy) [reply main.nf:514]: No you're right. If the stubs and shell produce a versions.yml file the this needs to be accessible. There are a few examples of this that needs to be checked.

File: main.nf (hunk 5 above)
[the remark this answers is not in the document]
Comment (ryanjameskennedy) [reply main.nf:584]: Is it a problem to specify it twice?

[the remark this answers is not in the document]
Comment (Jakob37) [reply main.nf:584]: No, my meaning here is that for some processed `path *versions.yml` have been specified. I think the purpose is to let Nextflow detect the output and forward it to the `publishDir` with the `*versions.yml` pattern.

In some cases only the `publishDir` has been specified, and not the `path` pattern. Will nextflow detect the version files in those cases?

[the remark this answers is not in the document]
Comment (Jakob37) [reply main.nf:584]: I.e. shouldn't it look like this:

```
process chanjo_sambamba {
	cpus 16
	memory '10 GB'
	publishDir "/${OUTDIR}/cov", mode: 'copy', overwrite: 'true', pattern: '*.cov'
	publishDir "/${OUTDIR}/versions", mode: 'copy' , overwrite: 'true', pattern: '*versions.yml'
	tag "$id"
	scratch true
	stageInMode 'copy'
	stageOutMode 'copy'

	when:
		params.varcall

	input:	
		set group, id, file(bam), file(bai) from chanjo_bam.mix(chanjo_bam_choice)

	output:
		file("${id}.bwa.chanjo.cov") into chanjocov
                path "*versions.yml"  <----- THIS LINE IS ADDED BY ME

```

**Pull request closed without merging**